mitmproxy/websocket.py

```diff
@@ -1,168 +1,136 @@
+    text and binary messages. To avoid a whole class of nasty type confusion bugs,
+    mitmproxy stores all message contents as binary. If you need text, you can decode the `content` property:
+
+    >>> if message.is_text:
+    >>>     text = message.content.decode()
     """
 
-    type: Opcode
-    """indicates either TEXT or BINARY (from wsproto.frame_protocol.Opcode)."""
     from_client: bool
     """True if this messages was sent by the client."""
-    content: Union[bytes, str]
+    is_text: bool
+    """
+    True if the message is a text message, False if the message is a binary message.
+
+    In either case, mitmproxy will store the message contents as *bytes*.
+    """
+    content: bytes
     """A byte-string representing the content of this message."""
     timestamp: float
     """Timestamp of when this message was received or created."""
 
-    killed: bool
-    """True if this messages was killed and should not be sent to the other endpoint."""
-
     def __init__(
         self,
-        type: int,
         from_client: bool,
-        content: Union[bytes, str],
+        is_text: bool,
+        content: bytes,
         timestamp: Optional[float] = None,
-        killed: bool = False
     ) -> None:
-        self.type = Opcode(type)  # type: ignore
         self.from_client = from_client
+        self.is_text = is_text
         self.content = content
         self.timestamp: float = timestamp or time.time()
-        self.killed = killed
 
     @classmethod
     def from_state(cls, state):
         return cls(*state)
 
     def get_state(self):
-        return int(self.type), self.from_client, self.content, self.timestamp, self.killed
+        return self.from_client, self.is_text, self.content, self.timestamp
 
     def set_state(self, state):
-        self.type, self.from_client, self.content, self.timestamp, self.killed = state
-        self.type = Opcode(self.type)  # replace enum with bare int
+        self.from_client, self.is_text, self.content, self.timestamp = state
 
     def __repr__(self):
-        if self.type == Opcode.TEXT:
-            return "text message: {}".format(repr(self.content))
+        if self.is_text:
+            return repr(self.content.decode(errors="replace"))
         else:
-            return "binary message: {}".format(strutils.bytes_to_escaped_str(self.content))
+            return repr(self.content)
 
     def kill(self):  # pragma: no cover
         """
         Kill this message.
 
-        It will not be sent to the other endpoint. This has no effect in streaming mode.
+        It will not be sent to the other endpoint.
         """
         warnings.warn(
             "WebSocketMessage.kill is deprecated, set an empty content instead.",
             DeprecationWarning,
             stacklevel=2,
         )
-        # empty str or empty bytes.
-        self.content = type(self.content)()
+        self.content = b""
 
+    @property
+    def killed(self) -> bool:  # pragma: no cover
```

Comment:
I only just realized that `kill` is also deprecated. I disagree with that decision, but I don't know what lead to that? Can you link to some discussion?

If I set the message to an empty string the original content is lost. In the UI it's just an empty message? But I want the user to know "Hey this message _would_ have been sent, but wasn't". In the same way we don't empty all HTTP headers and body if a flow is dropped. I don't want to lose any information.

Reply:
There is none, but there probably should have been some - I was just ripping out as much complexity as I could find. You mention a valid use case though. Going back to the drawing board, what would you expect `kill` to do? Promptly abort the connection, or just suppress the current message? I just now realize that we also don't have a good story for interception either.

Reply:
> Going back to the drawing board, what would you expect `kill` to do? Promptly abort the connection, or just suppress the current message? I just now realize that we also don't have a good story for interception either.

First of all we might want to call it `drop` and not introduce more violence :wink: 

Calling `messages[-1].drop()` should not forward the current message to the other end but keep it in the `messages` list with `dropped == True`.

This also touches another area that I always found awkward: when handling the `websocket_message` event I care about the message that this event is about. But it doesn't have a special place and I have to access the messages list. This feels weird to me and even though there probably can't be any race conditions with the current design (?) I always wonder if `messages[-1]` is actually guaranteed to be the message that triggered `websocket_message`.

Maybe we can pass the message as an additional parameter to the event? That would make for a much nicer API because that's what the user cares about.

Reply:
`drop` is an excellent suggestion. 👍 

Do we need separate `.drop()` and `.dropped`? Or does a single `.drop` property do the job?



I agree on your point regarding `messages[-1]`, even though there are no race conditions at the moment. :) I think adding an extra argument is good, that would also allow us to make `websocket_message` non-blocking, i.e. we can continue to handle messages while one is intercepted. But that is future work, let's get the basics down here.

Reply:
> Do we need separate `.drop()` and `.dropped`? Or does a single `.drop` property do the job?

I thought it should be consistent with `flow.kill()` but can you even detect if a flow has been killed currently? I just peeked into the flow.py and there is no flag?

To me `recent_message.drop = True` feels less like it actually does something than doing `recent_message.drop()`, which is an imperative. But I guess we already use that pattern with `flow.stream = True`. So maybe we should turn this around and update `flow.drop = True` in a future change and remove `flow.kill()`?

Here's one counter argument: if there is just a single `message.drop` property you don't know if it has already been dropped or not, it depends on the context. Inside of `websocket_message` it means that it will be dropped after the event (or all events) has been processed, but if you inspect any older message in the list then `drop == True` means it had been dropped. The benefit of a property (see streaming) is that multiple chained events can work on it.

So maybe just a `drop` property and an additional new `active` (or `live`?) so that you know exactly what's up.

Reply:
> But that is future work, let's get the basics down here.

Future work definitely references #3396 :wink: